src/Microsoft.AspNet.Hosting/HostingEngine.cs

```diff
@@ -29,18 +31,39 @@ public class HostingEngine
 
         public HostingEngine() : this(fallbackServices: null) { }
 
-        public HostingEngine(IServiceProvider fallbackServices)
+        public HostingEngine(IServiceProvider fallbackServices) : this(fallbackServices, applicationName: null, applicationPath: null) { }
+
+        public HostingEngine(IServiceProvider fallbackServices, string applicationName, string applicationPath)
         {
             _fallbackServices = fallbackServices ?? CallContextServiceLocator.Locator.ServiceProvider;
             _appLifetime = new ApplicationLifetime();
-            _applicationEnvironment = _fallbackServices.GetRequiredService<IApplicationEnvironment>();
-            _hostingEnvironment = new HostingEnvironment(_applicationEnvironment);
+            var appEnv = _fallbackServices.GetRequiredService<IApplicationEnvironment>();
+            _appEnv = new HostingApplicationEnvironment(appEnv, applicationPath, applicationName);
+            _hostingEnvironment = new HostingEnvironment(_appEnv);
             _fallbackServices = new WrappingServiceProvider(_fallbackServices, _hostingEnvironment, _appLifetime);
         }
 
+        public static IServiceProvider CreateApplicationServices(HostingContext context)
+        {
+            return CreateApplicationServices(context, fallbackServices: null);
+        }
+
+        // REVIEW: so ugly ApplicationName used here as input
```

Comment:
do you mean "This method exists for use in functional tests, not from product code"?